litebird_sim/beam_convolution.py

```diff
@@ -0,0 +1,359 @@
+# -*- encoding: utf-8 -*-
+
+import logging
+from dataclasses import dataclass
+from typing import Union, List, Dict, Optional
+
+import os
+
+import numpy as np
+import numpy.typing as npt
+from ducc0.totalconvolve import Interpolator, Interpolator_f
+
+from .coordinates import rotate_coordinates_e2g, CoordinateSystem
+from .hwp import HWP
+from .mueller_convolver import MuellerConvolver
+from .observations import Observation
+from .pointings import get_hwp_angle
+from .spherical_harmonics import SphericalHarmonics
+
+# Name of the environment variable used in the convolution
+NUM_THREADS_ENVVAR = "OMP_NUM_THREADS"
+
+
+@dataclass
+class BeamConvolutionParameters:
+    """Parameters used by the 4π beam convolution code
+
+    Fields:
+
+    - ``lmax`` (int): Maximum value for ℓ for the sky and beam coefficients
+    - ``mmax`` (int): Maximum value for m (azimuthal moment) for beam coefficients,
+      must be ≦ lmax - 4
+    - ``single_precision`` (bool): Set it to ``False`` to use 64-bit floating points
+      in the calculation
+    - ``epsilon`` (float): The desired relative accuracy of the interpolation
+    """
+
+    lmax: int
+    mmax: int
+    single_precision: bool = True
+    epsilon: float = 1e-5
+
+
+def add_convolved_sky_to_one_detector(
+    tod_det,
+    sky_alms_det: SphericalHarmonics,
+    beam_alms_det: SphericalHarmonics,
+    pointings_det,
+    mueller_matrix,
+    hwp_angle,
+    convolution_params: Optional[BeamConvolutionParameters] = None,
+    nthreads: int = 0,
+):
+    """ """
+
+    if not convolution_params:
+        sky_lmax = sky_alms_det.lmax
+
+        beam_lmax = beam_alms_det.lmax
+        beam_mmax = beam_alms_det.mmax
+
+        default_lmax = min(sky_lmax, beam_lmax)
+        default_mmax = min(default_lmax - 4, beam_mmax)
+
+        logging.warning(
+            (
+                "No convolution parameters, I will use the defaults "
+                "(ℓ_max={lmax}, m_max={mmax}), but this "
+                "might lead to unexpected errors and "
+                "gross misestimates"
+            ).format(lmax=default_lmax, mmax=default_mmax)
+        )
+        convolution_params = BeamConvolutionParameters(
+            lmax=default_lmax,
+            mmax=default_mmax,
+        )
+
+    else:
+        assert convolution_params.lmax - 4 >= convolution_params.mmax, (
+            "Error in the convolution parameters m_max must be ≦ ℓ_max!"
+            "Here ℓ_max={lmax} and m_max={mmax}!"
+        ).format(lmax=convolution_params.lmax, mmax=convolution_params.mmax)
```

Comment:
The assertion error is inconsistent with the text.

Reply:
Right